### DISCUSSION TITLE
Is there a way to make sure the same query with the same variables isn't executed too frequently?

### DISCUSSION THREAD

**Discussion by bkinsey808** [urql-graphql/urql#1914] [Q&A] [answered]
Consider the case where multiple components call the same query with the same variables at the same time. I want to make sure the server isn't bombarded with multiple requests, but instead the requests are throttled to be no more frequent per a certain interval of time.

**Comment by bkinsey808** [comment, marked as answer]
I wonder if https://github.com/FormidableLabs/urql/tree/main/exchanges/request-policy will do what i want.

**Comment by JoviDeCroock** [comment]
Queries using the same query-string and variables will be automatically deduped by the default `dedupExchange` you're using so you don't have to worry about that normally 😅 when the `operation.key` is the same it will share the observed stream, this `key` is generated from the query & variables.

**Reply by bkinsey808** [reply]
But, correct me if I'm wrong, dedupExchange only stops a query if another one is being fetched at the same time. 

from the docs: 
`An exchange that keeps track of ongoing Operations that haven't returned had a corresponding OperationResult yet. Any duplicate Operation that it receives is filtered out if the same Operation has already been received and is still waiting for a result.`

What I want is to block the query if the same query with the same variables has been run within X amount of time. Where X could be global across the entire app, that's fine. (Thanks for telling me about `key`. I looked a little bit at some exchange code, and I was wondering about that.)

Effectively, what I'm trying to do is query rate-limiting, so the server ppl don't get mad at me.

**Reply by JoviDeCroock** [reply]
Exactly yes, if you want to conditionally upgrade operations to cache and network to trigger a server hit then the requestPolicyExchange is your friend. Do note that as long as you use the default cache-first you only hit the server on the first query and after that the cache is updated by means of mutations/subscription-messages.
